peer_client.go

```diff
@@ -431,44 +421,46 @@ func (c *PeerClient) run() {
 				queue = nil
 
 				go func() {
-					ctx2 := tracing.StartNamedScope(ctx, "Send batch")
-					defer tracing.EndScope(ctx2, nil)
-					intervalSpan := trace.SpanFromContext(ctx2)
-					intervalSpan.SetAttributes(
-						attribute.String("batchWait", c.conf.Behavior.BatchWait.String()),
-					)
-
-					c.sendQueue(ctx2, queue2)
+					c.sendBatch(ctx, queue2)
 				}()
 			}
 		}
 	}
 }
 
-// sendQueue sends the queue provided and returns the responses to
+// sendBatch sends the queue provided and returns the responses to
 // waiting go routines
-func (c *PeerClient) sendQueue(ctx context.Context, queue []*request) {
-	ctx = tracing.StartScopeDebug(ctx)
+func (c *PeerClient) sendBatch(ctx context.Context, queue []*request) {
+	ctx = tracing.StartNamedScopeDebug(ctx, "PeerClient.sendBatch")
 	defer tracing.EndScope(ctx, nil)
-	span := trace.SpanFromContext(ctx)
-	span.SetAttributes(
-		attribute.Int("queueLen", len(queue)),
-		attribute.String("peer.grpcAddress", c.conf.Info.GRPCAddress),
-	)
 
 	batchSendTimer := prometheus.NewTimer(metricBatchSendDuration.WithLabelValues(c.conf.Info.GRPCAddress))
 	defer batchSendTimer.ObserveDuration()
-	funcTimer := prometheus.NewTimer(metricFuncTimeDuration.WithLabelValues("PeerClient.sendQueue"))
+	funcTimer := prometheus.NewTimer(metricFuncTimeDuration.WithLabelValues("PeerClient.sendBatch"))
 	defer funcTimer.ObserveDuration()
 
 	var req GetPeerRateLimitsReq
 	for _, r := range queue {
+		// NOTE: This trace has the same name because it's in a separate trace than the one above.
+		// We link the two traces, so we can relate our rate limit trace back to the above trace.
+		r.ctx = tracing.StartNamedScopeDebug(r.ctx, "PeerClient.sendBatch",
```

Comment:
Very interesting approach.  Now you can find this trace both in the upstream request and as background job.